ReadMe.md
# Course: CS 30
# Period: 2
# Date created: 19/09/21
# Date last modified: 28/10/21
# Name: Humzah Zahid Malik
# Description: Batman x Green Lantern - Temple Escape

changelog
# changelog
All notable changes of this project will be documented here 

## [0.0.4] 2021-11-25
### Added
- Exception handling has been added to the game for maps

## [0.0.4] 2021-11-02
### Added
- A new part of the game is added
- Map Class and Hero Class is added 

## [0.0.3] 2021-10-21
### Added
- A map is created and added to the code
- The map doesn't require movement. It only prints out.

### Changed
- All functions are stored into modules based on their type and category


## [0.0.2] 2021-09-30
### Added
- inventory is added to the game
- inventory consists of hints for each specific character
- A new action for each character is added as an hint

### Changed
- The original character is changed with two new characters
- The original storyline of the game is changed

### removed
- The original character is removed


## [0.0.1] 2021-09-16
# Original version
- A text-based adventure game is created
- consists of one character only
- It's purpose is to move around a "temple" and find his way out
- a quit option is added to exit the game at any point

maps.py
# This defines the layout of the temple
# Level 1 is where the user starts in the game
# Level 10 is where the game finishes


def map_temple():
    """ Prints out the map for the second part of the game """
    temple_map = [
      ["Blank", "Blank", "Blank", "Level 10", "Blank"],
      ["Blank", "Blank", "Blank", "Level 9", "Blank"],
      ["Blank", "Blank", "Level 7", "Level 8", "Blank"],
      ["Blank", "Blank", "Level 6", "Blank", "Blank"],
      ["Blank", "Level 4", "Level 5", "Blank", "Blank"],
      ["Blank", "Level 3", "Blank", "Blank", "Blank"],
      ["Blank", "Level 2", "Level 1", "Blank", "Blank"], ]
    print(temple_map)


def main_map():
    """ Prints out the map for the first part of the game """
    gamemap_2 = [
            ["seventhTile=(1, 2)", "eighthTile=(2, 2)", "ninthTile=(3, 2)", ],
            ["fourthTile=(1, 1)", "fifthTile=(2, 1)", "sixthTile=(3, 1)", ],
            ["firstTile=(1, 0)", "secondTile=(2, 0)", "thirdTile=(3, 0)", ],
                ]
    print("""All the locations of the map with their x-y coordinates
    are printed below:
    """)
    print(gamemap_2)

def main_map_2():
  try:
    main_map()
  except ImportError:
    print("There is an error with importing the file")
    main_map()

def main_map_3():
  try:
    map_temple()
  except ImportError:
    print("There is an error with importing the file")
    map_temple()

intro.py
def introduction_part_1():
    """ Prints out the introduction
    of the game """
    print("""
  Batman and Green Lantern are on a mission to find an underground enemy base.
  While they are looking for the base, the enemy is notified of their mission
  Batman and Green Lantern need to find their way out and escape the temple.

  There are two parts of the game.

  In the first part, you need to find the treasure,
  which is hidden in 1 of the 9 tiles.

  In the second part, after you have found the treausre,
  you will be automatically moved to the next part,
  where you will need to find your way out of the temple.

  ########################################################
  """)


def introduction_part_2():
    """ Prints out the introduction
    of the first part of the game """
    print("""
    THIS IS PART ONE OF THE GAME

    You have to find the treasure which is hidden
    in one of the tiles in the map.

    You can go only four directions to escape:
    forward, backward, right, left

    Here is a printed map of the temple.
    """)


def introduction_part_3():
    """ Prints out the introduction
    of the second part of the game """
    print("""
    THIS IS PART TWO OF THE GAME

    You can go only four directions to escape:
    forward, backward, right, left

    You also have an option to quit,
    which is accessible in the menu

    Here is a printed map of the temple. You can use this or a hint,
    provided in the menu, to find your way out of the temple:
   """)


Characters.py
class Character(object):
    """ This object gives attributes to each character """
    def __init__(self, name, special_power, birthplace):
        self.name = name
        self.special_power = special_power
        self.birthplace = birthplace
        self.location = 'start'

    def introduction(self):
        print('Hi, my name is ', self.name)
        print(""" I have been chosen by you for this mission.
        Here are some of my qualitites and background: """)

    def power_birthplace(self):
        print("My special power is my", self.special_power)
        print("My birthplace is", self.birthplace)
        print('\n')


def green_lantern_class():
    """ This object uses the main class to make a child object"""
    class Green_Lantern(Character):
        green_lantern = Character("Green Lantern", "power ring", "Coast City")
        green_lantern.introduction()
        green_lantern.power_birthplace()


def batman_class():
    """ This object uses the main class to make a child object"""
    class Batman(Character):
        batman = Character("Batman", "High Intelligence", "Gotham City")
        batman.introduction()
        batman.power_birthplace()


def Batman_inventory():
    """ Prints out a statement about Batman's inventory """
    print("This is Batman's inventory:")


def Green_lantern_inventory():
    """ Prints out a statement about Green Lantern's inventory """
    print("This is Green Lantern's inventory:")


def character_intro():
    """ Prints out the introduction for Batman """
    print("""
  ####################################################

  You have chosen Batman as your character.
  You will use Batman to get out of this maze.
  You can use a hint to complete a level.
  Now Batman will introduce himself.
  """)


def character_intro_2():
    """ Prints out the introduction for Green Lantern """
    print("""
    #####################################################

     You have chosen Green Lantern as your character.
     You will use Green Lantern to get out of this maze.
     You can use a hint to complete a level.
     Now Green Lantern will introduce himself.
    """)


main_inventory.py
# This inventory is paired with specific characters
inventory = {"Batman": {"Night Vision Goggles":
                        {"description": """ Use the Night Vision Goggles to see
                        in the dark and find your way """, }},
             "Green Lantern": {"Power Ring":
                               {"description":
                                "Use it as a flashlight to find their way", }}}

# This dictionary includes the descriptions of Batman


character = {"Batman":
             {"description":
              """I am a vigilante and my real name is Bruce Wayne!
              """}}

# This dictionary includes the description of Green Lantern


character_2 = {"Green Lantern":
               {"description":
                "My name is Hal Jordan. I come from the Planet Mogo"}}


def player_inventory(player, inventory):
    """Print out the inventory for the choosen character"""
    for item in inventory[player]:
        description = inventory[player][item]["description"]
        print(f"{player}'s {item} - {description}")


def character_inventory(player, character):
    """ prints out the description for Batman """
    for item in character[player]:
        description = character[player]["description"]
        print(f"{player}'s {item} - {description}")


def character_inventory_2(player_2, character_2):
    """ prints out the description for Green Lantern """
    for item in character_2[player_2]:
        description_2 = character_2[player_2]["description"]
        print(f"{player_2}'s {item} - {description_2}")


class_functions.py
import actions
import intro
import main_inventory
import maps

# This inventory is paired with specific characters
inventory = {"Batman": {"Night Vision Goggles":
                        {"description": """ Use the Night Vision Goggles to see
                        in the dark and find your way """, }},
             "Green Lantern": {"Power Ring":
                               {"description":
                                "Use it as a flashlight to find their way", }}}


class Character(object):
    """ This object gives attributes to each character """
    def __init__(self, name, special_power, birthplace):
        self.name = name
        self.special_power = special_power
        self.birthplace = birthplace
        self.location = 'start'

    def introduction(self):
        print('Hi, my name is ', self.name)
        print(""" I have been chosen by you for this mission.
        Here are some of my qualitites and background:""")

    def power_birthplace(self):
        print("My special power is my", self.special_power)
        print("My birthplace is", self.birthplace)
        print('\n')


def green_lantern_class():
    """ Prints out some description for Green Lantern """
    class Green_Lantern(Character):
        green_lantern = Character("Green Lantern", "power ring", "Coast City")
        green_lantern.introduction()
        green_lantern.power_birthplace()


def batman_class():
    """ Prints out some description for Batman """
    class Batman(Character):
        batman = Character("Batman", "High Intelligence", "Gotham City")
        batman.introduction()
        batman.power_birthplace()


class MapTile:
    """ This Map uses x and y coordinates to print out location"""
    def __init__(self, x, y):
        self.x = x
        self.y = y

    def location(self):
        print("your current location is ", self.x, self.y)

valid_actions = ["forward", "backward", "left", "right"]


def menu():
    """ Prints out the menu for the actions """
    print("""Choose an action:
     """)
    for action in valid_actions:
        print(f"* {action}")


def firstTiles():
    """Prints out the menu and the current location for movement in the map """
    main_inventory.player_inventory("Batman", inventory)
    class firstTile(MapTile):
        print("""
        ###################################################

        You are at the start. enter direction to go
        """)
        Tile = MapTile(1, 0)
        Tile.location()
    menu()
    direction = input("Where do you want to go: ")
    if direction.lower() == "forward":
            print(" You are going forward")
            fourthTiles()
    if direction.lower() == "backward":
            print("wrong way")
            firstTile()
    if direction.lower() == "left":
            print("wall ahead")
            firstTile()
    if direction.lower() == "right":
        print("You are going right to the second Tile")
        secondTiles()


def secondTiles():
    """Prints out the menu and the current location for movement in the map """
    class secondTile(MapTile):
        print("""
        You are at the second tile. enter direction to go
        """)
        Tile_2 = MapTile(2, 0)
        Tile_2.location()
        menu()
        direction = input("Where do you want to go: ")
        if direction.lower() == "forward":
            print("You are going forward to the fifth Tile")
            fifthTiles()
        if direction.lower() == "backward":
            print("dead end")
            secondTiles()
        if direction.lower() == "left":
            print("You are going right to the third Tile")
            firstTiles()
        if direction.lower() == "right":
            print("You are going right to the third Tile")
            thirdTiles()


def thirdTiles():
    """Prints out the menu and the current location for movement in the map """
    class thirdTile(MapTile):
        print("""
        You are at the third tile. enter direction to go
        """)
        Tile_2 = MapTile(3, 0)
        Tile_2.location()
        menu()
        direction = input("Where do you want to go: ")
        if direction.lower() == "forward":
            print("You are going forward to the sixth Tile")
            sixthTiles()
        if direction.lower() == "backward":
            print("Dead end")
            thirdTiles()
        if direction.lower() == "left":
            print("You are going left back to the second Tile")
            secondTiles()
        if direction.lower() == "right":
            print("Wall ahead")
            thirdTiles()


def fourthTiles():
    """Prints out the menu and the current location for movement in the map """
    class fourthTile(MapTile):
        print("""
        You are at the fourth tile. enter direction to go
        """)
        Tile_2 = MapTile(1, 1)
        Tile_2.location()
        menu()
        direction = input("Where do you want to go: ")
        if direction.lower() == "forward":
            print("You are going forward to the seventh Tile")
            seventhTiles()
        if direction.lower() == "backward":
            print("You are going back to the first Tile")
            firstTiles()
        if direction.lower() == "left":
            print("Dead End")
            fourthTiles()
        if direction.lower() == "right":
            print("You are going right to the fifth Tile")
            fifthTiles()


def fifthTiles():
    """Prints out the menu and the current location for movement in the map """
    class fifthTile(MapTile):
        print("""
        You are at the fifth tile. enter direction to go
        """)
        Tile_2 = MapTile(2, 1)
        Tile_2.location()
        menu()
        direction = input("Where do you want to go: ")
        if direction.lower() == "forward":
            print("You are going forward to the eighth Tile")
            eighthTiles()
        if direction.lower() == "backward":
            print("You are going back to the second Tile")
            secondTiles()
        if direction.lower() == "left":
            print("You are going left to the fourth Tile")
            fourthTiles()
        if direction.lower() == "right":
            print("You are going right to the sixth Tile")
            sixthTiles()


def sixthTiles():
    """Prints out the menu and the current location for movement in the map """
    class sixthTile(MapTile):
        print("""
        You are at the sixth tile. enter direction to go
        """)
        Tile_2 = MapTile(3, 1)
        Tile_2.location()
        menu()
        direction = input("Where do you want to go: ")
        if direction.lower() == "forward":
            print("You are going forward to the ninth Tile")
            ninthTiles()
        if direction.lower() == "backward":
            print("You are going back to the third Tile")
            thirdTiles()
        if direction.lower() == "left":
            print("You are going left to the fifth Tile")
            fifthTiles()
        if direction.lower() == "right":
            print("Dead End")
            sixthTiles()


def seventhTiles():
    """Prints out the menu and the current location for movement in the map """
    class seventhTile(MapTile):
        print("""
        You are at the seventh tile. enter direction to go
        """)
        Tile_2 = MapTile(1, 2)
        Tile_2.location()
        menu()
        direction = input("Where do you want to go: ")
        if direction.lower() == "forward":
            print("Wall ahead")
            seventhTiles()
        if direction.lower() == "backward":
            print("You are going back to the fourth Tile")
            fourthTiles()
        if direction.lower() == "left":
            print("Dead End")
            seventhTiles()
        if direction.lower() == "right":
            print("You are going right to the eighth Tile")
            eighthTiles()


def eighthTiles():
    """Prints out the menu and the current location for movement in the map """
    class eighthTile(MapTile):
        print("""
        You are at the eighth tile. enter direction to go
        """)
        Tile_2 = MapTile(2, 2)
        Tile_2.location()
        print(""" Congratulations!!!. You have found the treasure.
                  You will now move to the next round, where you have
                  to escape the enemies with the treasure. """)
        intro.introduction_part_3()
        maps.map_temple()
        actions.action_1()


def ninthTiles():
    """Prints out the menu and the current location for movement in the map """
    class ninthTile(MapTile):
        print("""
        You are at the ninth tile. enter direction to go
        """)
        Tile_2 = MapTile(3, 2)
        Tile_2.location()
        menu()
        direction = input("Where do you want to go: ")
        if direction.lower() == "forward":
            print("Wall ahead")
            ninthTiles()
        if direction.lower() == "backward":
            print("You are going back to the sixth Tile")
            sixthTiles()
        if direction.lower() == "left":
            print("You are going left to the eighth Tile")
            eighthTiles()
        if direction.lower() == "right":
            print("Dead end")
            ninthTiles()


game.py
# Course: CS 30
# Period: 2
# Date created: 19/09/21
# Date last modified: 21/09/21
# Name: Humzah Zahid Malik
# Description: Batman x Green Lantern - Temple Escape


import Characters
import os
import sys
import intro
import class_functions
import main_inventory
import maps

# This inventory is paired with specific characters
inventory = {"Batman": {"Night Vision Goggles":
                        {"description": """ Use the Night Vision Goggles to see
                        in the dark and find your way """, }},
             "Green Lantern": {"Power Ring":
                               {"description":
                                "Use it as a flashlight to find their way", }}}

character_choice = input("Enter Batman or Green Lantern: ")
if character_choice == "Batman":
    intro.introduction_part_1()
    intro.introduction_part_2()
    maps.main_map_2()
    Characters.character_intro()
    Characters.batman_class()
    Characters.Batman_inventory()
    class_functions.firstTiles()
    main_inventory.player_inventory("Batman", inventory)
if character_choice == "Green Lantern":
    intro.introduction_part_1()
    intro.introduction_part_2()
    maps.main_map()
    Characters.character_intro_2()
    Characters.green_lantern_class()
    Characters.Green_lantern_inventory()
    main_inventory.player_inventory("Green Lantern", inventory)
    class_functions.firstTiles()
